File: MenuFabrik/Models/Meal.swift
import Foundation
import SwiftData

enum MealTime: String, Codable, CaseIterable {
    case lunch = "Midi"
    case dinner = "Soir"
}

enum MealStatus: String, Codable, CaseIterable {
    case planned = "Prévu"
    case restaurant = "Restaurant"
    case absent = "Absent"
    case work = "Travail"
    case leftovers = "Restes"
    case shopping = "Courses"
    case skipped = "Sauté"
}

@Model
final class Meal {
    var id: UUID = UUID()
    var date: Date
    var type: MealTime
    var status: MealStatus
    var recipe: Recipe?
    @Relationship var selectedSideDishes: [SideDish]
    var menu: WeeklyMenu?
    @Relationship var attendees: [Participant]?
    
    init(date: Date, type: MealTime, status: MealStatus = .planned, recipe: Recipe? = nil, selectedSideDishes: [SideDish] = [], attendees: [Participant]? = []) {
        self.date = date
        self.type = type
        self.status = status
        self.recipe = recipe
        self.selectedSideDishes = selectedSideDishes
        self.attendees = attendees
    }
}

File: MenuFabrikWeb/src/models/Meal.ts
import type { Participant } from './Participant';
import type { Recipe } from './Recipe';
import type { SideDish } from './SideDish';

export const MealTime = {
    BREAKFAST: "Petit-déjeuner",
    LUNCH: "Midi",
    SNACK: "Collation",
    DINNER: "Soir",
    OTHER: "Autre"
} as const;
export type MealTime = typeof MealTime[keyof typeof MealTime];

export const MealStatus = {
    PLANNED: "Prévu",
    RESTAURANT: "Restaurant",
    ABSENT: "Absent",
    WORK: "Travail",
    LEFTOVERS: "Restes",
    SHOPPING: "Courses",
    SKIPPED: "Sauté"
} as const;
export type MealStatus = typeof MealStatus[keyof typeof MealStatus];

export interface Meal {
    id?: string;
    date: Date | string; // Firebase utilise des Timestamp, mais souvent on manipule des string ISO "2024-03-01T12:00:00Z"
    type: MealTime;
    status: MealStatus;

    // Pour différencier un repas contenant potentiellement une recette d'une simple note libre
    format?: 'standard' | 'note';
    noteText?: string;

    // Pour des questions de facilité avec Firestore en NoSQL, il est souvent préférable
    // d'embarquer (embed) une vue simplifiée de la recette ou juste son ID. 
    // Ici on garde l'objet pour l'algo, mais on met un ID en ref
    recipeId?: string;
    recipe?: Recipe;

    selectedSideDishIds: string[];
    selectedSideDishes?: SideDish[]; // Résolu via l'ID

    attendeeIds: string[];
    attendees?: Participant[]; // Résolu via l'ID

    // Ordre d'affichage dans la journée (pour permettre le déplacement manuel)
    order?: number;
}
